### PR TITLE
Allow dimension fields to have multiple values in standard and logsdb index mode

### LINKED ISSUE #112232: Fields with enabled time_series_dimension don't support multi values in non tsdb index modes
For OTel mappings, we're using a common definition of attributes and resource attributes for logs, metrics, and traces to avoid duplications and ensure a consistent mapping:

https://github.com/elastic/elasticsearch/blob/73c5c1e1c587cc7ec7ce1f0d10fea49ecfd39002/x-pack/plugin/otel-data/src/main/resources/component-templates/otel%40mappings.yaml#L20-L24

https://github.com/elastic/elasticsearch/blob/73c5c1e1c587cc7ec7ce1f0d10fea49ecfd39002/x-pack/plugin/otel-data/src/main/resources/component-templates/semconv-resource-to-ecs%40mappings.yaml#L9-L15

That's why we set `time_series_dimension` to `true` which I had expected to only kick in for TSDB data streams and ignored for others. However, it seems like a validation gets kicked in which disallows attributes to have multiple values (due to https://github.com/elastic/elasticsearch/issues/110387), even in standard and logsdb index modes. As a result, this exception is thrown:

https://github.com/elastic/elasticsearch/blob/73c5c1e1c587cc7ec7ce1f0d10fea49ecfd39002/server/src/main/java/org/elasticsearch/index/mapper/DocumentDimensions.java#L103

I think the reason for that is that even for `IndexMode.STANDARD`, we return an implementation of `DocumentDimensions` that only allows single values:

https://github.com/elastic/elasticsearch/blob/73c5c1e1c587cc7ec7ce1f0d10fea49ecfd39002/server/src/main/java/org/elasticsearch/index/IndexMode.java#L107-L109

Maybe we should return a noop implementation for index modes standard and tsdb? Or silently ignore `time_series_dimension` when the index mode is not `tsdb`.

We can also discuss other options for how we can compose our mappings so that multi-valued attributes are allowed for logs and traces while ensure consistency across the mappings and not having to duplicate mappings (in particular the ECS alias mappings, which are quite long)

cc @elastic/obs-ds-intake-services @gregkalapos

### LINKED ISSUE #112239: LogsDB - Ignore dimension fields
### Description

Multiple field types support the `dimension` boolean parameter after the introduction of TSDB (dimensions used for routing purposes). There are integrations sharing component and composable templates and creating index templates, with field mappings that are shared in case a filed is used both for metrics and logging integrations. As a result, it is possible that a mapping defining `dimension: true` on a field is used when index mode is `standard`, `time_series` or `logsdb`. We do not want to prevent this sharing to happen because that would lead to many more composable/component/index templates to maintain for integrations. As a result, the idea is to just ignore the `dimension` parameter when not applicable (`index.mode != `time_series`).

This happens specifically with Otel mappings, which use shared mappings for logs and metrics which means some fields are declared as dimensions just because those fields are also used when the index mode is `time_series`. Because of this, anyway, multi-valued dimension fields trigger a validation error when index mode is `logsdb` (and `standard`). Multi-valued fields being disallowed should only happen in the context of metrics anyway. See https://github.com/elastic/elasticsearch/issues/112232

We have different options to fix this:
* ignore the `dimension` attribute in mappers when index mode is not `time_series` (doing this is a bit painful for every mapper, the suggestion is to extract common behaviour in a parent class).
* always return empty dimensions when index mode is not `time_series`

NOTE: this can cause integrations to break when moving to LogsDB in case multi-value attributes are written.

### PR DESCRIPTION
**Pull request by felixbarny**
Fixes https://github.com/elastic/elasticsearch/issues/112232
Fixes https://github.com/elastic/elasticsearch/issues/112239

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentDimensions.java [hunk 1]
```diff
@@ -104,4 +104,47 @@ private void add(String fieldName) {
             }
         }
     };
+
+    /**
+     * Noop implementation that doesn't perform validations on dimension fields
+     */
+    enum Noop implements DocumentDimensions {
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentDimensions.java:111]: Lets' replace  `OnlySingleValueAllowed`, not needed afaict.

**Review by kkrik-es — APPROVED** [review]
Let's wait for Martijn to take a look too, in case we missed something.

File: server/src/test/java/org/elasticsearch/index/mapper/BooleanFieldMapperTests.java [hunk 2]
```diff
@@ -257,15 +259,24 @@ public void testDimensionIndexedAndDocvalues() {
...
     }
 
-    public void testDimensionMultiValuedField() throws IOException {
+    public void testDimensionMultiValuedField() throws Throwable {
         XContentBuilder mapping = fieldMapping(b -> {
             minimalMapping(b);
             b.field("time_series_dimension", true);
         });
-        DocumentMapper mapper = randomBoolean() ? createDocumentMapper(mapping) : createTimeSeriesModeDocumentMapper(mapping);
+        IndexMode indexMode = randomFrom(IndexMode.values());
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/index/mapper/BooleanFieldMapperTests.java:267]: Maybe not randomize `indexMode` here and in other similar tests? I would prefer to have one unit test testing the case where the validation kicks in and another whether the validation doesn't kick in (in that test randomizing between standard and logsdb does make sense).

This should avoid the situation where someone thinks tests passes and then only after re-running would figure out that a change doesn't make sense.

**Review by martijnvg — APPROVED** [review]
Left one testing comment, LGTM otherwise.

File: server/src/test/java/org/elasticsearch/index/mapper/BooleanFieldMapperTests.java (hunk 2 above)
Comment (felixbarny) [reply server/src/test/java/org/elasticsearch/index/mapper/BooleanFieldMapperTests.java]: Makes sense, I've adjusted the tests.

**Comment by felixbarny** [comment]
@elasticmachine test this please

**Pull request merged**